cpp/src/arrow/compute/kernels/scalar_nested.cc

```diff
@@ -252,6 +266,35 @@ struct StructFieldFunctor {
     return Status::OK();
   }
 
+  static Result<std::shared_ptr<Array>> ApplyFieldRef(KernelContext* ctx,
+                                                      const FieldRef& field_ref,
+                                                      std::shared_ptr<Array> current) {
+    if (current->type_id() != Type::STRUCT) {
+      return Status::Invalid("Not a StructArray: ", current->ToString(),
+                             "\nMaybe a bad FieldRef? ", field_ref.ToString());
+    }
+
+    if (field_ref.IsName()) {
+      const auto& array = checked_cast<const StructArray&>(*current);
+      current = array.GetFieldByName(*field_ref.name());
+      if (current == nullptr) {
+        return Status::Invalid("Field not found in struct: '", *field_ref.name(), "'");
+      }
+    } else if (field_ref.IsFieldPath()) {
+      for (const auto& idx : field_ref.field_path()->indices()) {
+        ARROW_RETURN_NOT_OK(CheckIndex(idx, *current->type()));
+        const auto& array = checked_cast<const StructArray&>(*current);
+        ARROW_ASSIGN_OR_RAISE(current, array.GetFlattenedField(idx, ctx->memory_pool()));
+      }
+    } else {
+      DCHECK(field_ref.IsNested());
+      for (const auto& ref : *field_ref.nested_refs()) {
+        ARROW_ASSIGN_OR_RAISE(current, ApplyFieldRef(ctx, ref, current));
+      }
```

Comment:
Possible alternative: convert the nested field ref to a FieldPath (eg with FieldRef::FindOne?) before the code to handle a FieldPath, and then it can use the same code.